Add AppDynamics for monitoring (#4410)
mendix/docs: content/en/docs/developerportal/operate/metrics/monitoring-with-saas/_index.md

> ---

Author: This document is now the landing page for Saas metrics. It has two child pages
appdynamics-metrics
datadog-metrics

This document has information that is relevant to both - mainly a description of the metrics

title: "Monitoring Your Mendix Apps with Saas"
linktitle: "Monitoring with Saas"
url: /developerportal/operate/monitoring-with-saas/
weight: 30
description: "How to use a Saas solution to monitor your Mendix Cloud v4 app."
tags: ["Datadog", "Mendix Cloud", "v4", "monitoring", "analysis", "Saas", "AppDynamics"]
---

## 1 Introduction

There are a number of monitoring and analysis tools for cloud applications which are available through a Software-as-a-service-based (Saas-based) data analytics platform. They provide comprehensive monitoring of servers, databases, tools, and services.

Mendix provides out-of-the-box configuration to use Datadog and AppDynamics to provide additional monitoring for your Mendix Apps running on the Mendix Cloud.

{{% alert color="info" %}}
Monitoring using a Saas solution is only available to apps running on Mendix Cloud v4.

> [AppDynamics](https://www.appdynamics.com/) logging and application metrics are supported in Mendix version 6.2 and above.

Reviewer: related https://github.com/mendix/docs/pull/4410/files#r842519793

Reviewer 2: For AppDynamics integration, we plan to provide only the [Application Monitoring](https://docs.appdynamics.com/21.4/en/application-monitoring/overview-of-application-monitoring) module. 
This includes :
- Java application metrics and business transactions via the [AppDynamics Java agent](https://docs.appdynamics.com/22.1/en/application-monitoring/install-app-server-agents/java-agent)
- Mendix runtime metrics via the [Machine Agent](https://docs.appdynamics.com/22.1/en/infrastructure-visibility/machine-agent/extensions-and-custom-metrics), which will be available as custom metrics.

Logging is part of the Logs Analytics module which won't be included.

Author: So: 
```suggestion
[AppDynamics](https://www.appdynamics.com/) application monitoring is supported in Mendix version 6.2 and above.
```

[Datadog](https://www.datadoghq.com/) logging and application metrics are supported in Mendix version 7.15 and above.

> AppDynamics and Datadog do not conflict, so you can use both if you need to.

Reviewer 2: Do we need to mention this explicitly? cc: @udh-mendix

Reviewer 3: For now there are some limitation for simultaneous usage. so I would suggest to remove this. Actually, after latest changes only basic functionality of AppDynamics is available while Datadog is active. In the future it very likely will be improved.

Author: Removed this.

* [Datadog for the Mendix Cloud](/developerportal/operate/datadog-metrics/)

A summary of the useful metrics provided by Mendix is in the section [Summary of Mendix Metrics](#summary), below.

{{% alert color="warning" %}}
Monitoring using a Saas is not supported in the deprecated Mendix Cloud v3, nor in default deployment buildpacks for other cloud platforms.
{{% /alert %}}

## 2 What Information Can Mendix Supply to a Saas-based Data Analytics Platform?

Mendix provides two types of monitoring data:

* Data from the Mendix Runtime, the Java Virtual Machine (JVM), the database, and the SaaS (Software as a Service, for example Cloud Foundry) environment – this is described in [Environment Metrics](#environment), below
* Data from within the Mendix app itself – this is described in [App Metrics](#app-metrics), below

## 3 Environment Metrics{#environment}

Once you have configured your data analytics platform for your app, it will automatically send information about the environment in which your app is running.

### 3.1 Environment Metrics Namespaces

The metrics from your app's environment are supplied in the following namespaces:

> * database – metrics on the database performance — by default, these are just *gauge* metrics (see [Rate and Count Database Metrics](#database-metrics), below)

Reviewer 2: The rate and count database metrics are Datadog specific. Will be available only when Datadog is enabled.

Author: But we will send some database metrics? So perhaps just say:
```suggestion
* database – metrics on the database performance
```
The stuff about rate and count metrics is in the datadog documentation anyway.

* jetty – metrics from the ingress controller of the app
* jmx – metrics from the Mendix runtime
* jvm – metrics from the Java virtual machine in which the Mendix runtime runs
* postgresql – database metrics specific to PostgreSQL databases
* system – metrics from the base system running on the platform or PaaS

### 3.2 Useful Metrics for Mendix Apps

The following metrics are useful when monitoring the performance of your Mendix app:

* mx.database.diskstorage_size
* jvm.heap_memory
* jvm.heap_memory_committed
* jvm.heap_memory_init
* jvm.heap_memory_max
* jvm.non_heap_memory
* jvm.non_heap_memory_committed
* jvm.non_heap_memory_init
* jvm.non_heap_memory_max
* postgresql.connections
* postgresql.database_size
* postgresql.max_connections
* postgresql.percent_usage_connections

> {{% todo %}}[What are Commons metrics - should they be documented here?]{{% /todo %}}

Author: Currently researching commons.pool2 metrics.

Author: Added to runtime metrics in https://github.com/mendix/docs/pull/4435

#### 4.1.2 Microflow Metrics

For the microflows you select (see [Customizing the Metrics Agent](#customizing), below), the metrics agent will provide the following metrics relating to the time the microflow takes to run:

* mx.microflow.time.avg
* mx.microflow.time.count
* mx.microflow.time.max
* mx.microflow.time.median
* mx.microflow.time.95percentile

#### 4.1.3 Activity Metrics

For the activities you select (see [Customizing the Metrics Agent](#customizing), below), the metrics agent will provide the following metrics:

* mx.activity.time.avg
* mx.activity.time.count
* mx.activity.time.max
* mx.activity.time.median
* mx.activity.time.95percentile

### 4.2 Customizing the Metrics Agent{#customizing}

By default, Mendix will pass a log of *all* **request handler** activity to your data analytics platform and no other information. However, by using JSON to configure the metrics agent, you can add logs of microflows and activities within microflows, and restrict which request handler calls are sent.

#### 4.2.1 Format of Metrics Agent Configuration{#mx-agent-format}

> <!-- Changes to this should be reflected in "Configuring the Java Instrumentation Agent" in /developerportal/deploy/private-cloud-monitor,

Author: This needs to be reviewed by the Mx4PC team to see whether anything needs to be changed in the Mx4PC monitoring document.

---
title: "Monitoring Your Mendix Apps with Saas"
linktitle: "Monitoring with Saas"
url: /developerportal/operate/monitoring-with-saas/
weight: 30
description: "How to use a Saas solution to monitor your Mendix Cloud v4 app."
tags: ["Datadog", "Mendix Cloud", "v4", "monitoring", "analysis", "Saas", "AppDynamics"]
---

## 1 Introduction

There are a number of monitoring and analysis tools for cloud applications which are available through a Software-as-a-service-based (Saas-based) data analytics platform. They provide comprehensive monitoring of servers, databases, tools, and services.

Mendix provides out-of-the-box configuration to use Datadog and AppDynamics to provide additional monitoring for your Mendix Apps running on the Mendix Cloud.

{{% alert color="info" %}}
Monitoring using a Saas solution is only available to apps running on Mendix Cloud v4.

> [AppDynamics](https://www.appdynamics.com/) application monitoring is supported in Mendix version 6.2 and above.

Reviewer 3: After testing we decided to specify that AppDynamics is fully supported in Mendix version 9.7 and later. For older versions (begin from 7.23) the limited amount of metrics is available on AppDynamics - 'postgres.' and 'mx.client.' Also the basic functionality of AppDynamics Java Agent is available. 
https://github.com/mendix/cf-mendix-buildpack/pull/508/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R421

Author: Updated

* Data from within the Mendix app itself – this is described in [App Metrics](#app-metrics), below

{{% alert color="info" %}}
If you are using AppDynamics with `APPDYNAMICS_MACHINE_AGENT_ENABLED` set to `false`, you will only receive the general AppDynamics metrics, provided by the AppDynamics Java Agent.

We recommend you set this to `true` to receive the analytics which will help you monitor your app environment.
{{% /alert %}}

## 3 Environment Metrics{#environment}

Once you have configured your data analytics platform for your app, it will automatically send information about the environment in which your app is running.

### 3.1 Environment Metrics Namespaces

The metrics from your app's environment are supplied in the following namespaces:

* commons.pool2 – database connection pool metrics
* database – metrics on the database performance
* jetty – metrics from the ingress controller of the app
* jmx – metrics from the Mendix runtime
* jvm – metrics from the Java virtual machine in which the Mendix runtime runs
* postgresql – database metrics specific to PostgreSQL databases

> * system – metrics from the base system running on the platform or PaaS

Reviewer 4: Hey @MarkvanMents, I know this part is a common part for all PaaS but for Datadog there is a special env var(`DD_ENABLE_CHECKS`) needed to enable the system metrics. In the end of the datadog-metrics.md there is a reference to Datadog's docs but that is even not mentioned there. The only place we mentioned this env var is our buildpack [readme](https://github.com/mendix/cf-mendix-buildpack). We are getting questions/tickets from the support regarding this sometimes, see https://mendix.slack.com/archives/C015ZNPFRPX/p1648217158171019 for example.

So I think we may want to add this information somewhere.

Author: Added
